File: src/com/lushprojects/circuitjs1/client/XMLDeserializer.java
/*    
    Copyright (C) Paul Falstad
    
    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

package com.lushprojects.circuitjs1.client;

import java.util.Vector;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.Random;
import java.lang.Math;

import com.google.gwt.xml.client.XMLParser;
import com.google.gwt.xml.client.Document;
import com.google.gwt.xml.client.Element;
import com.google.gwt.xml.client.NamedNodeMap;
import com.google.gwt.xml.client.Node;
import com.google.gwt.xml.client.NodeList;

class XMLDeserializer {

    CirSim app;

    XMLDeserializer(CirSim app_) { app = app_; }

    Element currentXmlElement;
    CircuitElm currentElm;

    void readCircuit(String text, int readFlags) {
        // Parse the XML
        Document doc = XMLParser.parse(text);
        
        // Get the root element
        Element root = doc.getDocumentElement();
        
	// Get all child elements of root
        NodeList children = root.getChildNodes();

	app.clearCircuit();
	currentXmlElement = root;

        int flags = parseIntAttr("f", 0);
	app.loader.readCircuitFlags(flags);
	SimulationManager sim = app.sim;
        sim.maxTimeStep = sim.timeStep = parseDoubleAttr("ts", sim.maxTimeStep);
        double sp = parseDoubleAttr("ic", app.getIterCount());
        int sp2 = (int) (Math.log(10*sp)*24+61.5);
        app.speedBar.setValue(sp2);
        app.currentBar.setValue(parseIntAttr("cb", app.currentBar.getValue()));
        CircuitElm.voltageRange = parseDoubleAttr("vr", CircuitElm.voltageRange);
	app.powerBar.setValue(parseIntAttr("pb", app.powerBar.getValue()));
        sim.minTimeStep = parseDoubleAttr("mts", sim.minTimeStep);
        app.setGrid();
        
        for (int i = 0; i < children.getLength(); i++) {
            Node node = children.item(i);
            if (node.getNodeType() != Node.ELEMENT_NODE)
		continue;

	    Element elem = (Element) node;
	    String tagName = elem.getTagName();
	    
	    if (tagName.equals("o")) {
		Scope sc = new Scope(app, app.sim);
		currentXmlElement = elem;
		sc.undumpXml(this);
		app.scopeManager.addScope(sc);
		continue;
	    }
	    if (tagName.equals("dm")) {
		currentXmlElement = elem;
		DiodeModel.undumpModelXml(this);
		continue;
	    }
	    if (tagName.equals("tm")) {
		currentXmlElement = elem;
		TransistorModel.undumpModelXml(this);
		continue;
	    }
	    if (tagName.equals("clm")) {
		currentXmlElement = elem;
		CustomLogicModel.undumpModelXml(this);
		continue;
	    }
	    if (tagName.equals("ccm")) {
		currentXmlElement = elem;
		CustomCompositeModel.undumpModelXml(this);
		continue;
	    }
	    
	    String x = elem.getAttribute("x");
	    String xs[] = x.split(" ");
	    int x1 = Integer.parseInt(xs[0]);
	    int y1 = Integer.parseInt(xs[1]);
	    int x2 = Integer.parseInt(xs[2]);
	    int y2 = Integer.parseInt(xs[3]);
	    String className = CirSim.xmlDumpTypeMap.get(tagName);
	    if (className == null) {
		app.console("unrecognized xml element: " + tagName);
		continue;
	    }
	    CircuitElm elm = app.constructElement(className, 0, 0);
	    currentXmlElement = elem;
	    currentElm = elm;
	    elm.undumpXml(this);
	    elm.setPosition(x1, y1, x2, y2);
	    app.elmList.add(elm);
	}
        app.needAnalyze();
    }

    public double parseDoubleAttr(String attr, double def) {
	String v = currentXmlElement.getAttribute(attr);
	if (v == null)
	    return def;
	return Double.parseDouble(v);
    }

    public int parseIntAttr(String attr, int def) {
	String v = currentXmlElement.getAttribute(attr);
	if (v == null)
	    return def;
	return Integer.parseInt(v);
    }
    
    public boolean parseBooleanAttr(String attr, boolean def) {
	String v = currentXmlElement.getAttribute(attr);
	if (v == null)
	    return def;
	return Boolean.parseBoolean(v);
    }
    
    public String parseStringAttr(String attr, String def) {
	String s = currentXmlElement.getAttribute(attr);
	if (s == null)
	    return def;
    	s = s.replace("&quot;", "\"")
             .replace("&apos;", "'")
             .replace("&lt;", "<")
             .replace("&gt;", ">")
             .replace("&amp;", "&");
	return s;
    }

    public String parseContents() {
	return currentXmlElement.getFirstChild().getNodeValue();
    }

    public List<Element> getChildElements() {
        List<Element> elements = new ArrayList<>();
        NodeList children = currentXmlElement.getChildNodes();
        
        for (int i = 0; i < children.getLength(); i++) {
            Node node = children.item(i);
            if (node.getNodeType() == Node.ELEMENT_NODE) {
                elements.add((Element) node);
            }
        }
        return elements;
    }

    public void parseChildElement(Element e) { currentXmlElement = e; }
}

File: XMLDeserializer.ts
/*
    Copyright (C) Paul Falstad

    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

import { CirSim } from "./CirSim";
import { CircuitElm } from "./CircuitElm";
import { CircuitLoader } from "./CircuitLoader";
import { DiodeModel } from "./DiodeModel";
import { RelayModel } from "./RelayModel";
import { TransistorModel } from "./TransistorModel";
import { CustomLogicModel } from "./CustomLogicModel";
import { Scope } from "./Scope";

export class XMLDeserializer {
    app: CirSim;
    currentXmlElement: Element;
    currentElm: CircuitElm | null = null;

    constructor(app: CirSim) { this.app = app; }

    readCircuit(text: string, readFlags: number): void {
        const doc = new DOMParser().parseFromString(text, "text/xml");
        const root = doc.documentElement;

        if ((readFlags & CircuitLoader.RC_RETAIN) === 0)
            this.app.clearCircuit();
        this.currentXmlElement = root;

        if ((readFlags & CircuitLoader.RC_RETAIN) === 0) {
            const flags = this.parseIntAttr("f", 0);
            this.app.loader.readCircuitFlags(flags);
            const sim = this.app.sim;
            sim.maxTimeStep = sim.timeStep = this.parseDoubleAttr("ts", sim.maxTimeStep);
            const sp  = this.parseDoubleAttr("ic", this.app.getIterCount());
            const sp2 = Math.trunc(Math.log(10 * sp) * 24 + 61.5);
            const ui  = this.app.ui;
            ui.speedBar.setValue(sp2);
            ui.currentBar.setValue(this.parseIntAttr("cb", ui.currentBar.getValue()));
            CircuitElm.voltageRange = this.parseDoubleAttr("vr", CircuitElm.voltageRange);
            ui.powerBar.setValue(this.parseIntAttr("pb", ui.powerBar.getValue()));
            sim.minTimeStep = this.parseDoubleAttr("mts", sim.minTimeStep);
            sim.solverType  = this.parseIntAttr("st", sim.solverType);
            this.app.setGrid();
        }

        this.readElements(root);
        this.app.loader.finishReadCircuit(readFlags);
    }

    // read elements from an already-parsed XML document (e.g. from CustomCompositeModel.elmDoc)
    readCircuitFromDoc(doc: Document): void {
        this.app.clearCircuit();
        this.readElements(doc.documentElement);
        this.app.loader.finishReadCircuit(0);
    }

    readElements(root: Element): void {
        const children = root.childNodes;

        for (let i = 0; i < children.length; i++) {
            const node = children[i];
            if (node.nodeType !== Node.ELEMENT_NODE)
                continue;

            const elem    = node as Element;
            const tagName = elem.tagName;

            if (tagName === "o") {
                const sc = new Scope(this.app, this.app.sim);
                this.currentXmlElement = elem;
                sc.undumpXml(this);
                this.app.scopeManager.addScope(sc);
                continue;
            }
            if (tagName === "dm") {
                this.currentXmlElement = elem;
                DiodeModel.undumpModelXml(this);
                continue;
            }
            if (tagName === "rlm") {
                this.currentXmlElement = elem;
                RelayModel.undumpModelXml(this);
                continue;
            }
            if (tagName === "tm") {
                this.currentXmlElement = elem;
                TransistorModel.undumpModelXml(this);
                continue;
            }
            if (tagName === "clm") {
                this.currentXmlElement = elem;
                CustomLogicModel.undumpModelXml(this);
                continue;
            }
            if (tagName === "ccm") {
                this.currentXmlElement = elem;
                // (window as any).CustomCompositeModel?.undumpModelXml(this); TODO
                continue;
            }
            if (tagName === "h") {
                this.currentXmlElement = elem;
                this.app.hintType  = this.parseIntAttr("t", -1);
                this.app.hintItem1 = this.parseIntAttr("i1", 0);
                this.app.hintItem2 = this.parseIntAttr("i2", 0);
                continue;
            }
            if (tagName === "adj") {
                this.currentXmlElement = elem;
                (window as any).Adjustable?.undumpXml(this, this.app);
                continue;
            }
            if (tagName === "test") {
                const tm = (window as any).TestManager?.theManager;
                if (tm != null) {
                    this.currentXmlElement = elem;
                    tm.saveTestTag(this.parseDoubleAttr("len", -1));
                }
                continue;
            }
            if (tagName === "switchevent") {
                const tm = (window as any).TestManager?.theManager;
                if (tm != null) {
                    this.currentXmlElement = elem;
                    tm.saveSwitchEvent(this.parseDoubleAttr("t", 0), this.parseIntAttr("en", -1));
                }
                continue;
            }
            if (tagName === "scopedata") {
                const tm = (window as any).TestManager?.theManager;
                if (tm != null) {
                    this.currentXmlElement = elem;
                    const si     = this.parseIntAttr("si", -1);
                    const pi     = this.parseIntAttr("pi", -1);
                    const en     = this.parseIntAttr("en", -1);
                    const v      = this.parseIntAttr("v", 0);
                    const u      = this.parseIntAttr("u", 0);
                    const sp     = this.parseIntAttr("sp", 1);
                    const ts     = this.parseDoubleAttr("ts", 0);
                    const values = this.parseDoubleArray(this.parseContents());
                    tm.saveScopeDataTag(si, pi, en, v, u, sp, ts, values);
                }
                continue;
            }

            const x = elem.getAttribute("x");
            if (x == null)
                continue;
            const className = CirSim.xmlDumpTypeMap.get(tagName);
            if (className == null) {
                this.app.console("unrecognized xml element: " + tagName);
                continue;
            }
            const elm = this.app.constructElement(className, 0, 0);
            this.currentXmlElement = elem;
            this.currentElm = elm;
            elm.undumpXml(this);
            elm.setPositionFromXml(elem);
            this.app.elmList.push(elm);
        }
    }

    parseDoubleAttr(attr: string, def: number): number {
        const v = this.currentXmlElement.getAttribute(attr);
        if (v == null)
            return def;
        return parseFloat(v);
    }

    parseIntAttr(attr: string, def: number): number {
        const v = this.currentXmlElement.getAttribute(attr);
        if (v == null)
            return def;
        return parseInt(v);
    }

    parseBooleanAttr(attr: string, def: boolean): boolean {
        const v = this.currentXmlElement.getAttribute(attr);
        if (v == null)
            return def;
        return v === "true";
    }

    parseStringAttr(attr: string, def: string | null): string | null {
        let s = this.currentXmlElement.getAttribute(attr);
        if (s == null)
            return def;
        s = s.replace(/&quot;/g, '"')
             .replace(/&apos;/g, "'")
             .replace(/&lt;/g, "<")
             .replace(/&gt;/g, ">")
             .replace(/&amp;/g, "&");
        return s;
    }

    parseContents(): string | null {
        try {
            const child = this.currentXmlElement.firstChild;
            return child ? child.nodeValue : null;
        } catch (e) {
            return null;
        }
    }

    parseDoubleArray(text: string | null): number[] {
        if (text == null || text.trim() === "")
            return [];
        return text.trim().split(",").map(s => parseFloat(s.trim()));
    }

    getChildElements(): Element[] {
        const elements: Element[] = [];
        const children = this.currentXmlElement.childNodes;
        for (let i = 0; i < children.length; i++) {
            const node = children[i];
            if (node.nodeType === Node.ELEMENT_NODE)
                elements.push(node as Element);
        }
        return elements;
    }

    parseChildElement(elem: Element): void { this.currentXmlElement = elem; }
}
